File: src/codes/excessive-branching/py/rfc_state.py
from abc import ABC, abstractmethod

class DiscountState(ABC):
    @abstractmethod
    def calculate(self, base_price: float) -> float:
        pass

class StudentState(DiscountState):
    def calculate(self, base_price: float) -> float:
        return base_price * 0.8

class MemberState(DiscountState):
    def calculate(self, base_price: float) -> float:
        return base_price * 0.9

class VIPState(DiscountState):
    def calculate(self, base_price: float) -> float:
        return base_price * 0.7

class DefaultState(DiscountState):
    def calculate(self, base_price: float) -> float:
        return base_price

class DiscountContext:
    def __init__(self, state: DiscountState):
        self._state = state

    def set_state(self, state: DiscountState):
        self._state = state

    def calculate(self, base_price: float) -> float:
        return self._state.calculate(base_price)

# 利用例
context = DiscountContext(StudentState())
print(context.calculate(1000))  # 800

File: src/codes/excessive-branching/ts/rfc_state.ts
// 状態インターフェース
interface DiscountState {
  calculate(basePrice: number): number;
}

// 各状態クラス
class StudentState implements DiscountState {
  calculate(basePrice: number): number {
    return basePrice * 0.8;
  }
}

class MemberState implements DiscountState {
  calculate(basePrice: number): number {
    return basePrice * 0.9;
  }
}

class VIPState implements DiscountState {
  calculate(basePrice: number): number {
    return basePrice * 0.7;
  }
}

class DefaultState implements DiscountState {
  calculate(basePrice: number): number {
    return basePrice;
  }
}

// コンテキスト
class DiscountContext {
  private state: DiscountState;

  constructor(state: DiscountState) {
    this.state = state;
  }

  setState(state: DiscountState) {
    this.state = state;
  }

  calculate(basePrice: number): number {
    return this.state.calculate(basePrice);
  }
}

// 利用例
const context = new DiscountContext(new StudentState());
console.log(context.calculate(1000)); // 800
